### PR TITLE
Implemented code suggestions for save and load in PyTorch

### PR DESCRIPTION
**Pull request by Anilith**
Hello dr. Lu,

I made a separate pull request for the simple save/load functionality in PyTorch. I implemented the code suggestions you posted and checked the formatting using black.
I removed the epochs and the loss as you said they're not used, I checked the new implementation of ```self.opt.load_state_dict```and it seems to work.

Kind regards,
Olivier

### CODE REVIEW

File: deepxde/model.py [hunk 1]
```diff
@@ -593,33 +593,54 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
-        if backend_name != "tensorflow.compat.v1":
+            if protocol == "tf.train.Saver":
+                self.saver.save(
+                    self.sess, save_path, global_step=self.train_state.epoch
+                )
+            elif protocol == "pickle":
+                with open(
+                    "{}-{}.pkl".format(save_path, self.train_state.epoch), "wb"
+                ) as f:
+                    pickle.dump(self.state_dict(), f)
```
Comment (lululxvi) [inline deepxde/model.py:613]: The protocol "pickle" is for all backends. It should be right after the print info

File: deepxde/model.py [hunk 2]
```diff
@@ -593,33 +593,54 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
                 If `protocol` is "pickle", save using the Python pickle module. Only
                 "tf.train.Saver" protocol supports ``restore()``.
```
Comment (lululxvi) [inline deepxde/model.py:594]: The `protocol` argument should be "backend" or "pickle". If it is "backend", then use backend specific method. So the docstring should be:
```
protocol (string): If `protocol` is "backend", save using the backend-specific method.
     For "tensorflow.compat.v1", use `tf.train.Save <https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Saver#attributes>`_.
    For "pytorch", use `torch.save <https://pytorch.org/docs/stable/generated/torch.save.html>`_.
    If `protocol` is "pickle", save using the Python pickle module.
    Only the protocol "backend" supports ``restore()``.
```

File: deepxde/model.py [hunk 3]
```diff
@@ -593,33 +593,54 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
+            torch.save(
+                {
+                    "model_state_dict": self.net.state_dict(),
+                    "optimizer_state_dict": self.opt.state_dict(),
+                },
+                save_path + "{}.pt".format(self.train_state.epoch),
+            )
+        else:
             raise NotImplementedError(
                 "state_dict hasn't been implemented for this backend."
```
Comment (lululxvi) [inline deepxde/model.py:626]: "Model.save() hasn't been implemented for this backend."

File: deepxde/model.py [hunk 4]
```diff
@@ -593,33 +593,54 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
-        self.saver.restore(self.sess, save_path)
+        if backend_name == "tensorflow.compat.v1":
+            self.saver.restore(self.sess, save_path)
+        elif backend_name == "pytorch":
+            checkpoint = torch.load(save_path)
+            self.net.load_state_dict(checkpoint["model_state_dict"])
+            self.opt.load_state_dict(checkpoint["optimizer_state_dict"])
+        else:
+            raise NotImplementedError(
+                "state_dict hasn't been implemented for this backend."
```
Comment (lululxvi) [inline deepxde/model.py:642]: "Model.restore() hasn't been implemented for this backend."

File: deepxde/model.py [hunk 5]
```diff
@@ -593,33 +593,54 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
+                    pickle.dump(self.state_dict(), f)
+
+        elif backend_name == "pytorch":
+            # torch.save does not make a directory, so directory has to be created before training
+            torch.save(
+                {
+                    "model_state_dict": self.net.state_dict(),
+                    "optimizer_state_dict": self.opt.state_dict(),
+                },
+                save_path + "{}.pt".format(self.train_state.epoch),
```
Comment (lululxvi) [inline deepxde/model.py:622]: "-{}.pt"

File: deepxde/model.py [hunk 6]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
+    def save(self, save_path, protocol="backend", verbose=0):
         """Saves all variables to a disk file.
 
         Args:
-            protocol (string): If `protocol` is "tf.train.Saver", save using
-                `tf.train.Save <https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Saver#attributes>`_.
-                If `protocol` is "pickle", save using the Python pickle module. Only
-                "tf.train.Saver" protocol supports ``restore()``.
+            protocol (string): If `protocol` is "backend", save using the backend-specific method.
+            For "tensorflow.compat.v1", use `tf.train.Save <https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Saver#attributes>`_.
```
Comment (lululxvi) [inline deepxde/model.py:596]: You need indent of 4 spaces here.

File: deepxde/model.py [hunk 7]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
-        # TODO: backend tensorflow, pytorch
-        if backend_name != "tensorflow.compat.v1":
+        # TODO: backend tensorflow
+        if backend_name == "tensorflow.compat.v1":
+            destination = OrderedDict()
+            variables_names = [v.name for v in tf.global_variables()]
+            values = self.sess.run(variables_names)
+            for k, v in zip(variables_names, values):
+                destination[k] = v
+        if backend_name == "pytorch":
```
Comment (lululxvi) [inline deepxde/model.py:582]: elif

File: deepxde/model.py [hunk 8]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
+            destination = self.net.state_dict()
+
```
Comment (lululxvi) [inline deepxde/model.py:584]: no empty line here

File: deepxde/model.py [hunk 9]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
+        else:
             raise NotImplementedError(
                 "state_dict hasn't been implemented for this backend."
             )
-        destination = OrderedDict()
-        variables_names = [v.name for v in tf.global_variables()]
-        values = self.sess.run(variables_names)
-        for k, v in zip(variables_names, values):
-            destination[k] = v
         return destination
```
Comment (lululxvi) [inline deepxde/model.py:589]:
```python
        if backend_name == "tensorflow.compat.v1":
            ...
            return destination
        if backend_name == "pytorch":
            return self.net.state_dict()
        raise NotImplementedError(
            "state_dict hasn't been implemented for this backend."
        )
```

File: deepxde/model.py (hunk 7 above)
Comment (lululxvi) [inline deepxde/model.py:582]: elif

File: deepxde/model.py (hunk 8 above)
Comment (lululxvi) [inline deepxde/model.py:584]: no empty line

File: deepxde/model.py [hunk 10]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
-            self.saver.save(self.sess, save_path, global_step=self.train_state.epoch)
-        elif protocol == "pickle":
+        if protocol == "pickle":
             with open("{}-{}.pkl".format(save_path, self.train_state.epoch), "wb") as f:
                 pickle.dump(self.state_dict(), f)
 
+        elif protocol == "backend":
+            if backend_name == "tensorflow.compat.v1":
+
+                if protocol == "tf.train.Saver":
```
Comment (lululxvi) [inline deepxde/model.py:615]: Remove `if protocol == "tf.train.Saver":`

File: deepxde/model.py [hunk 11]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
+                    self.saver.save(
+                        self.sess, save_path, global_step=self.train_state.epoch
+                    )
+
+            elif backend_name == "pytorch":
+                # torch.save does not make a directory, so directory has to be created before training
```
Comment (lululxvi) [inline deepxde/model.py:621]: what does it mean by "before training"?

File: deepxde/model.py [hunk 12]
```diff
@@ -572,54 +572,80 @@ def op(inputs):
...
                 pickle.dump(self.state_dict(), f)
 
+        elif protocol == "backend":
+            if backend_name == "tensorflow.compat.v1":
+
+                if protocol == "tf.train.Saver":
+                    self.saver.save(
+                        self.sess, save_path, global_step=self.train_state.epoch
+                    )
+
```
Comment (lululxvi) [inline deepxde/model.py:619]: No empty line
